File: php-sdk/src/Receipt.php
<?php

namespace Plasm;

/**
 * Execution receipt proving work was done.
 *
 * As of phase-core M7 the canonical Phase receipt is
 * `SignedReceipt<JobResult>` with fields:
 *
 *     {
 *       "schema_version": 1,
 *       "result":         { ... JobResult ... },
 *       "job_id":         "hex-32-bytes",
 *       "worker_pubkey":  "hex-32-bytes",
 *       "signature":      "hex-64-bytes",
 *       "completed_at":   "2026-05-27T..."
 *     }
 *
 * The legacy November-2025 shape used by the local-transport mock path is:
 *
 *     {
 *       "version":      "0.1",
 *       "module_hash":  "sha256:...",
 *       "exit_code":    0,
 *       "wall_time_ms": 8,
 *       "timestamp":    1779896995,
 *       "node_pubkey":  "...",
 *       "signature":    "..."
 *     }
 *
 * This class accepts both. `isSignedEnvelope()` tells the caller which
 * shape is in play; the convenience getters expose the union of both shapes.
 */
class Receipt
{
    // ---- M7 SignedReceipt<JobResult> fields ----
    private ?int $schemaVersion = null;
    /** @var mixed Decoded JobResult body (or null for legacy receipts). */
    private $result = null;
    private string $jobId = '';
    private string $workerPubkey = '';
    private string $completedAt = '';

    // ---- legacy daemon-receipt fields (still used by LocalTransport mock) ----
    private string $version = '';
    private string $moduleHash = '';
    private int $exitCode = 0;
    private int $wallTimeMs = 0;
    private int $timestamp = 0;
    private string $nodePubkey = '';

    // ---- shared ----
    private string $signature = '';

    /** True when this receipt was loaded as an M7 SignedReceipt<JobResult>. */
    private bool $signedEnvelope = false;

    private function __construct(array $data)
    {
        // Detect format. The M7 envelope has both `schema_version` and `result`.
        if (array_key_exists('schema_version', $data)
            && array_key_exists('result', $data)) {
            $this->signedEnvelope = true;
            $this->schemaVersion = (int) $data['schema_version'];
            $this->result        = $data['result'];
            $this->jobId         = (string) ($data['job_id'] ?? '');
            $this->workerPubkey  = (string) ($data['worker_pubkey'] ?? '');
            $this->signature     = (string) ($data['signature'] ?? '');
            $this->completedAt   = (string) ($data['completed_at'] ?? '');

            // Surface the most commonly-needed JobResult fields on the receipt
            // so existing example code (`$receipt->getModuleHash()` etc.) keeps
            // working without forcing every caller through `getResult()`.
            $result = is_array($this->result) ? $this->result : [];
            $this->moduleHash = (string) ($result['metrics']['extra']['module_hash'] ?? '');
            $this->exitCode   = (int) ($result['metrics']['extra']['exit_code'] ?? 0);
            $this->wallTimeMs = (int) ($result['metrics']['total_duration_ms'] ?? 0);
        } else {
            // Legacy shape — no envelope.
            $this->signedEnvelope = false;
            $this->version      = (string) ($data['version'] ?? '');
            $this->moduleHash   = (string) ($data['module_hash'] ?? '');
            $this->exitCode     = (int) ($data['exit_code'] ?? 0);
            $this->wallTimeMs   = (int) ($data['wall_time_ms'] ?? 0);
            $this->timestamp    = (int) ($data['timestamp'] ?? 0);
            $this->nodePubkey   = (string) ($data['node_pubkey'] ?? '');
            $this->signature    = (string) ($data['signature'] ?? '');
        }
    }

    /**
     * Create a receipt from a JSON string. Auto-detects M7 vs legacy shape.
     */
    public static function fromJson(string $json): self
    {
        $data = json_decode($json, true);
        if (json_last_error() !== JSON_ERROR_NONE) {
            throw new \InvalidArgumentException(
                'Invalid JSON: ' . json_last_error_msg()
            );
        }
        return new self(is_array($data) ? $data : []);
    }

    /**
     * Create a legacy "mock" receipt for local execution. Used by
     * `LocalTransport` where there's no signed envelope to verify.
     */
    public static function createMock(string $moduleHash, int $exitCode, int $wallTimeMs): self
    {
        return new self([
            'version'      => '0.1',
            'module_hash'  => $moduleHash,
            'exit_code'    => $exitCode,
            'wall_time_ms' => $wallTimeMs,
            'timestamp'    => time(),
            'node_pubkey'  => 'local_execution',
            'signature'    => 'unsigned',
        ]);
    }

    /**
     * Serialize back to JSON in whichever shape this receipt was built from.
     */
    public function toJson(): string
    {
        if ($this->signedEnvelope) {
            return json_encode([
                'schema_version' => $this->schemaVersion,
                'result'         => $this->result,
                'job_id'         => $this->jobId,
                'worker_pubkey'  => $this->workerPubkey,
                'signature'      => $this->signature,
                'completed_at'   => $this->completedAt,
            ], JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES);
        }
        return json_encode([
            'version'      => $this->version,
            'module_hash'  => $this->moduleHash,
            'exit_code'    => $this->exitCode,
            'wall_time_ms' => $this->wallTimeMs,
            'timestamp'    => $this->timestamp,
            'node_pubkey'  => $this->nodePubkey,
            'signature'    => $this->signature,
        ], JSON_PRETTY_PRINT);
    }

    /**
     * Verify the receipt signature against a PINNED, caller-supplied public
     * key.
     *
     * Security contract (SEC-03 / audit C1+C3):
     *
     *  - The pinned key is MANDATORY. There is deliberately no zero-argument
     *    form. A verifier must already know the public key of the worker it
     *    trusts (from an operator allowlist, a libp2p PeerId binding, a
     *    pre-shared key, etc.) and pass it here. The key embedded in the
     *    receipt is attacker-controlled and is NEVER consulted for a trust
     *    decision — it may only be displayed via `getNodePubkey()` to show
     *    *who claims* to have signed.
     *
     *  - Only the M7 `phase-receipt:v1:` Ed25519-over-canonical-JSON path is
     *    trusted. Legacy (November-2025 pipe-format) receipts can still be
     *    parsed for display, but `verify()` returns false for them: there is
     *    no downgrade path to a weaker signing format.
     *
     *  - There is no `local_execution` (or any other) magic-string bypass.
     *    Trust in a locally-executed job must come from the transport context
     *    the caller controls, never from a field inside the untrusted object.
     *
     * @param string $expectedPubkeyHex Hex-encoded Ed25519 public key the
     *                                  caller trusts. Required.
     * @return bool True only if this is a v1 signed envelope whose signature
     *              verifies against `$expectedPubkeyHex`.
     */
    public function verify(string $expectedPubkeyHex): bool
    {
        // Legacy / non-envelope receipts are never trusted. No downgrade.
        if (!$this->signedEnvelope) {
            return false;
        }

        // Reject schema versions newer than we understand (mirrors the Rust
        // verifier, which errors on schema_version > SCHEMA_VERSION).
        if (($this->schemaVersion ?? 0) > Crypto::SCHEMA_VERSION) {
            return false;
        }

        if ($expectedPubkeyHex === '') {
            return false;
        }

        // Verify the v1 signature against the PINNED key only. The receipt's
        // own embedded worker_pubkey is intentionally ignored here.
        $message = Crypto::getCanonicalMessage($this);
        return Crypto::verifySignature($message, $this->signature, $expectedPubkeyHex);
    }

    // ------------------------------------------------------------------
    // Format-disambiguation
    // ------------------------------------------------------------------

    public function isSignedEnvelope(): bool { return $this->signedEnvelope; }
    public function getSchemaVersion(): int { return $this->schemaVersion ?? 0; }
    public function getResult() { return $this->result; }
    public function getCompletedAt(): string { return $this->completedAt; }
    public function getWorkerPubkey(): string { return $this->workerPubkey; }

    // ------------------------------------------------------------------
    // Compatibility getters — work for both shapes where possible
    // ------------------------------------------------------------------

    public function getVersion(): string
    {
        if ($this->signedEnvelope) {
            return 'phase-receipt:v' . ($this->schemaVersion ?? 1);
        }
        return $this->version;
    }

    public function getModuleHash(): string { return $this->moduleHash; }
    public function getExitCode(): int { return $this->exitCode; }
    public function getWallTimeMs(): int { return $this->wallTimeMs; }

    public function getTimestamp(): int
    {
        if ($this->signedEnvelope) {
            // M7 envelope carries an ISO-8601 timestamp; surface a Unix epoch
            // for callers that expect the legacy field shape.
            $t = strtotime($this->completedAt);
            return $t === false ? 0 : $t;
        }
        return $this->timestamp;
    }

    /**
     * The node / worker public key, hex-encoded. Returns `worker_pubkey`
     * for M7 envelopes and `node_pubkey` for legacy receipts.
     *
     * DISPLAY ONLY. This value comes straight from the untrusted receipt and
     * states only *who claims* to have signed. Never use it as the pinned key
     * for `verify()` — doing so would make verification self-referential and
     * meaningless (see audit C1).
     */
    public function getNodePubkey(): string
    {
        return $this->signedEnvelope ? $this->workerPubkey : $this->nodePubkey;
    }

    public function getSignature(): string { return $this->signature; }

    public function getJobId(): string
    {
        if ($this->signedEnvelope) {
            return $this->jobId;
        }
        return $this->moduleHash;
    }

    /** Alias retained for the remote_test.php example. */
    public function getNodeId(): string
    {
        return $this->getNodePubkey();
    }

    public function isSuccess(): bool { return $this->exitCode === 0; }
}

File: daemon/src/wasm/receipt.rs
use serde::{Deserialize, Serialize};

/// Execution receipt proving work was done
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Receipt {
    /// Receipt version
    pub version: String,

    /// SHA-256 hash of the executed WASM module
    pub module_hash: String,

    /// Exit code (0 = success)
    pub exit_code: u32,

    /// Wall clock execution time (milliseconds)
    pub wall_time_ms: u64,

    /// Timestamp (Unix epoch seconds)
    pub timestamp: u64,

    /// Node public key (Ed25519, hex-encoded)
    pub node_pubkey: String,

    /// Signature over receipt fields (Ed25519, hex-encoded)
    pub signature: String,
}

impl Receipt {
    /// Create a new unsigned receipt
    pub fn new(module_hash: String, exit_code: u32, wall_time_ms: u64) -> Self {
        use std::time::{SystemTime, UNIX_EPOCH};

        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs();

        Self {
            version: "0.1".to_string(),
            module_hash,
            exit_code,
            wall_time_ms,
            timestamp,
            node_pubkey: String::new(), // Set when signing
            signature: String::new(),    // Set when signing
        }
    }

    /// Sign the receipt with an Ed25519 private key
    pub fn sign(&mut self, _private_key: &[u8]) -> Result<(), String> {
        // TODO: Implement signing in later task
        // For now, return placeholder
        self.node_pubkey = "placeholder_pubkey".to_string();
        self.signature = "placeholder_signature".to_string();
        Ok(())
    }

    /// Verify the receipt signature
    pub fn verify(&self, _public_key: &[u8]) -> Result<bool, String> {
        // TODO: Implement verification in later task
        Ok(true)
    }

    /// Serialize to JSON
    pub fn to_json(&self) -> Result<String, String> {
        serde_json::to_string_pretty(self)
            .map_err(|e| format!("Failed to serialize receipt: {}", e))
    }

    /// Deserialize from JSON
    pub fn from_json(json: &str) -> Result<Self, String> {
        serde_json::from_str(json)
            .map_err(|e| format!("Failed to deserialize receipt: {}", e))
    }

    /// Load from JSON file
    pub fn from_file(path: &std::path::Path) -> Result<Self, String> {
        let content = std::fs::read_to_string(path)
            .map_err(|e| format!("Failed to read file: {}", e))?;
        Self::from_json(&content)
    }

    /// Save to JSON file
    pub fn to_file(&self, path: &std::path::Path) -> Result<(), String> {
        let json = self.to_json()?;
        std::fs::write(path, json)
            .map_err(|e| format!("Failed to write file: {}", e))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_receipt_creation() {
        let receipt = Receipt::new("sha256:abc123".to_string(), 0, 1500);
        assert_eq!(receipt.version, "0.1");
        assert_eq!(receipt.exit_code, 0);
        assert_eq!(receipt.wall_time_ms, 1500);
        assert!(receipt.timestamp > 0);
    }

    #[test]
    fn test_receipt_json_serialization() {
        let mut receipt = Receipt::new("sha256:abc123".to_string(), 0, 1500);
        receipt.sign(&[]).unwrap(); // Add placeholder signature

        let json = receipt.to_json().unwrap();

        // Verify JSON contains required fields
        assert!(json.contains("\"version\":"));
        assert!(json.contains("\"module_hash\":"));
        assert!(json.contains("\"signature\":"));

        // Deserialize and verify
        let loaded = Receipt::from_json(&json).unwrap();
        assert_eq!(loaded.module_hash, receipt.module_hash);
        assert_eq!(loaded.exit_code, receipt.exit_code);
        assert_eq!(loaded.wall_time_ms, receipt.wall_time_ms);
    }
}
